types/mdast/index.d.ts

```diff
@@ -10,52 +10,162 @@ export type AlignType = 'left' | 'right' | 'center' | null;
 
 export type ReferenceType = 'shortcut' | 'collapsed' | 'full';
 
-export type Content =
-    | TopLevelContent
-    | ListContent
-    | TableContent
-    | RowContent
-    | PhrasingContent;
-
-export type TopLevelContent =
-    | BlockContent
-    | FrontmatterContent
-    | DefinitionContent;
-
-export type BlockContent =
-    | Paragraph
-    | Heading
-    | ThematicBreak
-    | Blockquote
-    | List
-    | Table
-    | HTML
-    | Code;
+/**
+ * This map registers all node types that may be used where markdown block content is accepted.
+ *
+ * These types are accepted inside block quotes, list items, footnotes, and roots.
+ *
+ * This interface can be augmented to register custom node types.
+ *
+ * @example
+ * declare module 'mdast' {
+ *   interface BlockContentMap {
+ *     // Allow using toml nodes defined by remark-frontmatter.
+ *     toml: TOML;
+ *   }
+ * }
+ */
+export interface BlockContentMap {
+    paragraph: Paragraph;
+    heading: Heading;
+    thematicbreak: ThematicBreak;
+    blockquote: Blockquote;
+    list: List;
+    table: Table;
+    html: HTML;
+    code: Code;
+}
+
+/**
+ * This map registers all node definition types.
+ *
+ * This interface can be augmented to register custom node types.
+ *
+ * @example
+ * declare module 'mdast' {
+ *   interface DefinitionContentMap {
+ *     // Allow using toml nodes defined by remark-frontmatter.
+ *     toml: TOML;
+ *   }
+ * }
```

Comment:
:thinking: this may be outside the scope of this PR but something worth considering.
Let's say `remark-frontmatter` adds this snippet to it's typings.
Let's also say I have a package in my dependency tree that uses `remark-frontmatter`, but I don't directly import it, use it, or have access to the `TOML` type.

Would this pattern mean that, because a dependency uses `remark-frontmatter` I now have all the frontmatter node types added to my `mdast` types? :thinking:

Reply:
This is rather important to figure out, because I am planning on adding such integrations to the remark plugins!

Reply:
remark-frontmatter’s nodes are arbitrary: based on options. So could we add those nodes “programatically”? 🤔

Reply:
The augmentation happens if the **TypeScript** file in which it happens is imported. The JavaScript output is irrelevant.

Let’s say a package uses `remark-frontmatter`. If it uses the types by `remark-frontmatter` in its output, then `remark-frontmatter/index.d.ts` will be imported and the augmentations in it will get applied.

For example, the following plugin named `foo` will use `remark-frontmatter`, bus has a different default.

```ts
import { Plugin } from 'unified'
import frontmatter, { RemarkFrontmatterOptions } from 'remark-frontmatter'

export const foo: Plugin<[RemarkFrontmatterOptions]> = (options = ['toml']) => {
  return frontmatter(options)
}
```

This will generate the following type definitions in `node_modules/foo/index.d.ts`:

```ts
import { Plugin } from 'unified'
import { RemarkFrontmatterOptions } from 'remark-frontmatter'

export const foo: Plugin<[RemarkFrontmatterOptions]>;
```

Meaning if `foo` gets imported, `remark-frontmatter` also gets imported and its module augmentations get applied.

However, this example limits the use of `remark-frontmatter` by not allowing options:

```ts
import { Plugin } from 'unified'
import frontmatter from 'remark-frontmatter'

export const foo: Plugin<[]> = () => {
  return frontmatter(['toml'])
}
```

This will generate the following type definitions in `node_modules/foo/index.d.ts`:

```ts
import { Plugin } from 'unified'

export const foo: Plugin<[]>;
```

Meaning `remark-frontmatter` won’t be imported, so the augmentation doesn’t take place.

I think this is rare though.

---

> remark-frontmatter’s nodes are arbitrary: based on options. So could we add those nodes “programatically”? :thinking:

This can’t be done. If one wants to define new node types using `remark-frontmatter`, they need to define this manually.

```ts
import { Root, Literal } from 'mdast'
import * as remark from 'remark'
import * as frontmatter from 'remark-frontmatter'

interface Metal extends Literal {
  type: 'metal';
}

declare module 'mdast' {
  interface BlockContentMap {
    metal: Metal;
  }
}

function metalProcessor() {
  return (root: Root) => {
    for (const child of root.children) {
      if (child.type === 'metal') {
        // $ExpectType Metal
        child;
        child.type = 'text'
        child.value = `🤘${child.value}🤘`
      }
    }
  }
}

remark()
  .use(frontmatter, [ { type: 'metal', fence: '\m/' } ] )
  .use(metalProcessor)
```

What could be done is narrow down the types accepted by `remark-frontmatter` so it only accepts registered types.

Reply:
Thanks for the clarification @remcohaszing! :bow: 
My main concern was augmentation happening without an explicit import, it sounds like this will not be the case.